subsys/fs/nvs/nvs.c

```diff
@@ -822,105 +772,62 @@ int nvs_delete(struct nvs_fs *fs, u16_t id)
 	return nvs_write(fs, id, NULL, 0);
 }
 
-ssize_t nvs_read(struct nvs_fs *fs, u16_t id, void *data, size_t len)
-{
-	int rc;
-	struct nvs_entry entry;
-
-	if ((id == NVS_ID_EMPTY) || (id == NVS_ID_SECTOR_END)) {
-		return -EINVAL;
-	}
-	entry.id = id;
-	/* Read last entry */
-	rc = nvs_get_last_entry(fs, &entry);
-	if (entry.len == 0) {
-		return -ENOENT;
-	}
-	if (rc) {
-		goto err;
-	}
-	rc = nvs_check_crc(fs, &entry);
-	if (rc) {
-		goto err;
-	}
-	rc = nvs_flash_read(fs, entry.data_addr, data,
-			    min(entry.len, len));
-	if (rc) {
-		goto err;
-	}
-	rc = entry.len;
-
-err:
-	return rc;
-}
-
 ssize_t nvs_read_hist(struct nvs_fs *fs, u16_t id, void *data, size_t len,
 		  u16_t cnt)
 {
 	int rc;
-	struct nvs_entry entry;
+	u32_t wlk_addr, rd_addr;
 	u16_t cnt_his;
+	struct nvs_ate wlk_ate;
+	size_t ate_size;
 
-	if ((id == NVS_ID_EMPTY) || (id == NVS_ID_SECTOR_END)) {
+	ate_size = _nvs_al_size(fs, sizeof(struct nvs_ate));
+
+	if ((id == NVS_ID_EMPTY) || (len > (fs->sector_size - 2*ate_size))) {
 		return -EINVAL;
 	}
-	entry.id = id;
-	/* Read history entry */
-	/* First find out how many entries are in the history */
-	rc = nvs_get_first_entry(fs, &entry);
-	if (rc) {
-		goto err;
-	}
+
 	cnt_his = 0;
-	while (1) {
-		rc = nvs_walk_entry(fs, &entry);
+
+	wlk_addr = fs->ate_wra;
+	rd_addr = wlk_addr;
+
+	while (cnt_his <= cnt) {
+		rd_addr = wlk_addr;
+		rc = _nvs_prev_ate(fs, &wlk_addr, &wlk_ate);
 		if (rc) {
-			break;
+			goto err;
 		}
-		/* disregard items with zero length */
-		if (entry.len > 0) {
+		if ((wlk_ate.id == id) &&  (!_nvs_ate_crc8_check(&wlk_ate))) {
 			cnt_his++;
 		}
-	}
-	if (cnt_his < cnt) {
-		/* Trying to read history item that is not available */
-		return -ENOENT;
-	}
-	/* Now get the correct item by decreasing cnt_his until cnt
-	 * is reached
-	 */
-	rc = nvs_get_first_entry(fs, &entry);
-	if (rc) {
-		goto err;
-	}
-	while (1) {
-		if (cnt_his == cnt) {
+		if (wlk_addr == fs->ate_wra) {
 			break;
 		}
-		rc = nvs_walk_entry(fs, &entry);
-		if (rc) {
-			break;
-		}
-		/* disregard items with zero length */
-		if (entry.len > 0) {
-			cnt_his--;
-		}
 	}
 
-	if (rc) {
-		goto err;
-	}
-	rc = nvs_check_crc(fs, &entry);
-	if (rc) {
-		goto err;
+	if (((wlk_addr == fs->ate_wra) && (wlk_ate.id != id)) ||
+	    (wlk_ate.len == 0) || (cnt_his < cnt)) {
+		return -ENOENT;
 	}
-	rc = nvs_flash_read(fs, entry.data_addr, data,
-			    min(entry.len, len));
+
+	rd_addr &= 0xffff0000;
+	rd_addr += wlk_ate.offset;
+	rc = _nvs_flash_rd(fs, rd_addr, data, min(len, wlk_ate.len));
 	if (rc) {
 		goto err;
 	}
-	rc = entry.len;
+
+	return wlk_ate.len;
```

Comment:
In API documentation pleas rephrase `Any other value, indicates an error.` -> `Any negative value, indicates an error.`
This is applicable to the nvs_read() as well.

Reply:
This is fixed.